Samples/AudioPreprocessing/AudioPreprocessing/Model/PreprocessModel.cs

```diff
@@ -1,16 +1,323 @@
+
+        }
+        public ImageSource GenerateMelSpectrogram(string audioPath)
+        {
+            var signal = GetSignalFromFile(audioPath);
+            int batchSize = 1;
+            int signalSize = signal.Count;
+            int windowSize = 256;
+            int dftSize = 256;
+            int hopSize = 3;
+            int nMelBins = 1024;
+            int samplingRate = 8192;
+            //Bitmap bitmap = new Bitmap("C:\\Users\\t-janedu\\source\\repos\\Windows-Machine-Learning\\Samples\\AudioPreprocessing\\AudioPreprocessing\\tmp\\mel_spectrogram_3tones.jpg");
+
+
+            var rawSoftwareBitmap = GetMelpsectrogramFromSignal(signal, batchSize, signalSize, windowSize, dftSize,
+                    hopSize, nMelBins, samplingRate);
+            var bitmap = SoftwareToBitmap(rawSoftwareBitmap).Result;            
+            var imageSource = BitmapToImageSource(bitmap);
+            return imageSource;    
+        }
+        private async Task<Bitmap> SoftwareToBitmap(SoftwareBitmap softwareBitmap)
+        {
+            Bitmap bmp;
+            using (var stream = new InMemoryRandomAccessStream())
+            {
+                var encoder = await Windows.Graphics.Imaging.BitmapEncoder.CreateAsync(
+                    Windows.Graphics.Imaging.BitmapEncoder.PngEncoderId, stream);
+                // Set the software bitmap
+                encoder.SetSoftwareBitmap(softwareBitmap);
+                Console.WriteLine(stream.Size);
+                await encoder.FlushAsync();
+                bmp = new Bitmap(stream.AsStream());
+                return bmp;
+            }
+        }
+
+        private BitmapImage BitmapToImageSource(Bitmap bitmap)
+        {
+            using (MemoryStream memory = new MemoryStream())
+            {
+                bitmap.Save(memory, ImageFormat.Bmp);
+                memory.Position = 0;
+                BitmapImage bitmapimage = new BitmapImage();
+                bitmapimage.BeginInit();
+                bitmapimage.StreamSource = memory;
+                bitmapimage.CacheOption = BitmapCacheOption.OnLoad;
+                bitmapimage.EndInit();
+                return bitmapimage;
+            }
+        }
+
+        private List<float> Resample(List<float> signal, int oldSampleRate, int newSampleRate)
+        {
+            int nStep = oldSampleRate / newSampleRate;
+            var result = new List<float>();
+            for (int i = 0; i < signal.Count; i += nStep)
+            {
+                result.Add(signal[i]);
+            }
+            return result;
+        }
+
+
+        private List<float> GetSignalFromFile(string filename, int sampleRate = 8000, int amplitude = 5000)
+        {
+            List<float> signal = new List<float>();
+            int rawSampleRate;
+            using (var reader = new AudioFileReader(filename))
+            {
+                rawSampleRate = reader.WaveFormat.SampleRate;
```

Comment:
The entire file can be read into memory with reader.Read(buffer, offset, count)

So i believe this can be simplified to using an array for the signal with size reader.Length, and have it populated with:
reader.Read(buffer, offset, count).

Maybe the amplitude change can be done in the resampling code...

Reply:
As it turns out, reader.Length returns something kind of strange. It gives 64000, but the `reader.WaveFormat.SampleRate` is 16000 and the SpeechCommand files are 1 second long, which leads to a spectrogram like this:

![image](https://user-images.githubusercontent.com/14916563/124027354-c2cb8c80-d9c0-11eb-9e51-9623aecbbaa1.png)

Most of the extra "samples" are just 0's. I ended up cropping the array.

Since resampling would ideally be done with operators, I removed the Resample method from this file. I think it would make sense to not do the amplitude multiplying here, but is it ok to leave it until resampling is implemented in the graph?

Reply:
Fixed with reader.Length/ sizeof(float)